va/va.go

```diff
@@ -673,49 +685,157 @@ func (va *ValidationAuthorityImpl) performRemoteValidation(ctx context.Context,
-				// represents an internal error. The real error has already been logged
-				// so return a server internal problem result without detail.
-				result <- probs.ServerInternal("Remote PerformValidation RPCs failed")
+
+		// Store the first non-nil problem to return later (if `MultiVAFullResults`
+		// is enabled).
+		if firstProb == nil && prob != nil {
+			firstProb = prob
+		}
+
+		// If MultiVAFullResults isn't enabled then return early whenever the
+		// success or failure threshold is met.
+		if !features.Enabled(features.MultiVAFullResults) {
+			if good >= required {
+				state = "success"
+				return nil
+			} else if bad > va.maxRemoteFailures {
+				return prob
 			}
+		}
+
+		// If we haven't returned early because of MultiVAFullResults being enabled
+		// we need to break the loop once all of the VAs have returned a result.
+		if len(remoteProbs) == len(va.remoteVAs) {
+			break
+		}
+	}
+
+	// If we are using `features.MultiVAFullResults` then we haven't returned
+	// early and can now log the differential between what the primary VA saw and
+	// what all of the remote VAs saw.
+	va.logRemoteValidationDifferentials(domain, primaryResult, remoteProbs)
+
+	// Based on the threshold of good/bad return nil or a problem.
+	if good >= required {
+		state = "success"
+		return nil
+	} else if bad > va.maxRemoteFailures {
+		return firstProb
+	}
+
+	// This condition should not occur - it indicates the good/bad counts didn't
+	// meet either the required threshold or the maxRemoteFailures threshold.
+	return probs.ServerInternal("Too few remote PerformValidation RPC results")
+}
+
+// logRemoteValidationDifferentials is called by `processRemoteResults` when the
+// `MultiVAFullResults` feature flag is enabled. It produces a JSON log line
+// that contains the primary VA result and the results each remote VA returned.
+func (va *ValidationAuthorityImpl) logRemoteValidationDifferentials(
+	domain string,
+	primaryResult *probs.ProblemDetails,
+	remoteProbs []*probs.ProblemDetails) {
+
+	allEqual := true
+	for _, e := range remoteProbs {
+		if e != primaryResult {
+			allEqual = false
 			break
 		}
 	}
+	if allEqual {
+		// There's no point logging a differential line if the primary VA and remote
+		// VAs all agree.
+		return
+	}
+
+	logOb := struct {
+		Domain        string
+		PrimaryResult *probs.ProblemDetails
+		RemoteResults []*probs.ProblemDetails
+	}{
+		Domain:        domain,
+		PrimaryResult: primaryResult,
+		RemoteResults: remoteProbs,
+	}
 
-	va.metrics.remoteValidationTime.With(prometheus.Labels{
-		"type":   string(challenge.Type),
-		"result": state,
-	}).Observe(va.clk.Since(s).Seconds())
+	logJSON, err := json.Marshal(logOb)
+	if err != nil {
+		// log a warning - a marshaling failure isn't expected given the data and
+		// isn't critical enough to break validation for by returning an error to
+		// the caller.
+		va.log.Warningf("Could not marshal log object in "+
```

Comment:
Probably want a `return` here otherwise we'll log `remoteVADifferentials JSON=` after.

Reply:
Good catch, thanks. Fixed in 4f5131b